Add reccuring Postgres connection check for advisory lock

antekresic opened the pull request:
In a disaster or failover scenario, advisory lock that was previously
acquired would be lost by the leader instance. In this case, we need to
double check that the connection is down and release the
application-side lock to avoid having a state with multiple leaders.

feikesteenbergen reviewed (approved):
LGTM, although instead of specific checkConnection one, I would immediately check the lock at line 98, instead of `checkConnection`, just call `pg_try_advisory_lock` then and there, you could then remove the whole of line 117-128.

niksajakovljevic commented on pkg/util/lock.go:
```diff
@@ -108,7 +126,14 @@ func (l *PgAdvisoryLock) TryLock() (bool, error) {
-		log.Debug("msg", fmt.Sprintf("Lock obtained for group id %d", l.groupLockId))
+		log.Debug("msg", fmt.Sprintf("Lock obtained for group id %d", l.groupLockID))
+		l.lockTimer = time.AfterFunc(l.lockTimeout, func() {
+			result, _ := l.TryLock()
```
I think if a connection pooler (eg. pg-bouncer) is sitting between an adapter and PostgreSQL, we might fail to detect that we actually lost the lock (the connection might still show as valid to the adapter). I'd recommend that after we validate the connection we also try getting the lock from the db.

niksajakovljevic commented on cmd/prometheus-postgresql-adapter/main.go:
```diff
@@ -155,6 +156,7 @@ func parseFlags() *config {
 	flag.DurationVar(&cfg.prometheusTimeout, "leader-election-pg-advisory-lock-prometheus-timeout", -1, "Adapter will resign if there are no requests from Prometheus within a given timeout (0 means no timeout). "+
 		"Note: make sure that only one Prometheus instance talks to the adapter. Timeout value should be co-related with Prometheus scrape interval but add enough `slack` to prevent random flips.")
 	flag.BoolVar(&cfg.restElection, "leader-election-rest", false, "Enable REST interface for the leader election")
+	flag.DurationVar(&cfg.lockTimeout, "pg-advisory-lock-timeout", 3*time.Second, "PG advisory lock connection timeout duration. Adapter will resign in case of a lost connection since the lock will be lost as well.")
```
Nit: I'd recommend the following argument description: `An interval used by adapter leader to make sure we still hold the lock`

niksajakovljevic replied:
Look at my comment about using existing `ScheduledElector` and not introducing new Timer.

niksajakovljevic commented on pkg/util/lock.go:
```diff
@@ -108,7 +126,14 @@ func (l *PgAdvisoryLock) TryLock() (bool, error) {
 	if l.obtained {
-		log.Debug("msg", fmt.Sprintf("Lock obtained for group id %d", l.groupLockId))
+		log.Debug("msg", fmt.Sprintf("Lock obtained for group id %d", l.groupLockID))
+		l.lockTimer = time.AfterFunc(l.lockTimeout, func() {
```
I'd recommend not creating another timer, but instead relying on existing `ScheduledElector`. `ScheduledElector` is responsible for running leader check on certain interval . I guess we would just need to change `IsLeader` implementation of PG advisory lock to actually validate the connection and try re-obtaining the lock. In this case election interval would need to be configurable and default to something more reasonable. (eg. 5 seconds).

niksajakovljevic reviewed (changes requested):
Overall looks good. My major concern is properly handling a use-case when there is a connection pooler in between and some code structure.

antekresic committed:
Add reccuring Postgres connection check for advisory lock

In a disaster or failover scenario, advisory lock that was previously
acquired would be lost by the leader instance. In this case, we need to
double check that the connection is down and release the
application-side lock to avoid having a state with multiple leaders.

niksajakovljevic reviewed (approved):
Looks good!

RobAtticus commented on cmd/prometheus-postgresql-adapter/main.go:
```diff
@@ -193,42 +195,41 @@ func buildClients(cfg *config) (writer, reader) {
 	}
+	if cfg.prometheusTimeout == -1 {
+		log.Error("msg", "Prometheus timeout configuration must be set when using PG advisory lock")
+		os.Exit(1)
```
perhaps in a follow up PR: maybe change this function to return an error, and then have the caller handle the error with `os.Exit(1)`
